docs/proposal/20210823-gpu-support.md

```diff
@@ -0,0 +1,143 @@
+
+## Motivation
+
+CAPV does not currently support GPU-accelerated workloads which have a limited addressable market. Competitively, all of the hyper-scale cloud providers (AWS, Azure, & GCP) offer GPU-accelerated virtualization and Kubernetes platforms. Many of them additionally offer robust vertically integrated AI/ML solutions building on open source technology. GPU support is a long-requested feature. With this capability, users can leverage GPU support.
+
+### Goals
+
+1. To enable GPU Support on Vsphere Kubernetes Cluster.
+2. To allow vGPU support for multi-node clusters.
+3. To allow GPU Direct support for multi-node clusters.
+4. To allow PCI passthrough support for single-node clusters.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - using vGPU
+
+Stacy works at an IT department of an organization where AI/ML models are run every hour and the entire staff wants to leverage GPU-supported nodes. VGPU's are able to provide close to bare-metal performance. Stacy needs a lot of sharing GPUs.
+vGPU provides a full-fleshed cluster from which a virtual GPU can be requested.
+
+#### Story 2 - GPU Direct implementation
+
+Tony is an Engineer at a big organization with multiple centers across the Globe. Tony needs a big GPU cluster to address the needs of his organization. So they intend to use GPU Direct. GPU direct is technically a pool of GPUs connected by a network card in a kind of peer-to-peer network. This is the latest technology and gives the advantage of shared GPUs from a pool. GPU  direct is the technology that helps in linking multiple GPUs. Tony wants to leverage this technology so that it is easier for him to manage resource allocation.
+
+#### Story 3 - PCI passthrough for single node Customer
+
+Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.
+
+### Requirements
+
+#### Functional Requirements
+
+* FR1: CAPV MUST support vGPUs.
+* FR2: CAPV MUST support GPUDirect.
+* FR3: CAPV MUST support PCI Passthrough.
+
+#### Non-Functional Requirements
+
+* NFR1: Unit tests MUST exist for all 3 supports
+* NFR2: e2e tests MUST exist for 3 supports
+
+### Implementation Details/Notes/Constraints
+
+#### Controller Changes
+
+```go
+// vGPUProfileName is the profileName of the vGPU, in string.
+// Defaults to the eponymous property value in the template from which the
+// virtual machine is cloned.
+// +optional
+VGPUProfileName string `json:"vGPUProfileName,omitempty"`
```

Comment:
Missing some info on how this string will get integrated into the existing API types. Also, in order to support PCI passthrough, we'd need more information. From the [docs](https://vdc-download.vmware.com/vmwb-repository/dcr-public/b50dcbbf-051d-4204-a3e7-e1b618c1e384/538cf2ec-b34f-4bae-a332-3820ef9e7773/vim.vm.device.VirtualPCIPassthrough.AllowedDevice.html) the `deviceId` and `vendorId` are required

Reply:
As we are talking about vGPU, GPU direct also, these needs to covered in the schema and possibly in a generic way. What are your thoughts? 
According to the [docs](https://blogs.vmware.com/apps/2018/09/using-gpus-with-virtual-machines-on-vsphere-part-3-installing-the-nvidia-grid-technology.html): 
```
A vGPU profile allows you to assign a GPU solely to one virtual machine’s use or to be used in a shared mode with others.
```
That means `VGPUProfileName ` does not fit with other mode i.e. `pci-passthrough` or `gpu-direct`

Reply:
@sonasingh46 i think we need to define different schema for different techniques, as defined in vm-operator https://github.com/vmware-tanzu/vm-operator/blob/main/config/crd/bases/vmoperator.vmware.com_virtualmachineclasses.yaml#L31-L38
Not sure though, this would be applicable only when each mode has different fields.

Reply:
>i think we need to define different schema for different techniques
Sure.  
I meant is those schemas needs to be present in the CR spec. ( The bigger CR schema ) 
Something like this https://github.com/vmware-tanzu/vm-operator-api/blob/master/api/v1alpha1/virtualmachineclass_types.go#L55

Also, while we are designing the schema, it definitely helps to draw inspiration from the vm-operator api types and it is possible that we might come up with slight different approach for the schema for GPU after discussions. 
Open to discussions.

Reply:
I have removed 'v' from the variable names as this was not Vgpu specific.